Spokestack/TFLiteKeywordRecognizer.swift

```diff
@@ -0,0 +1,373 @@
+    
+    /// Global state for the speech pipeline.
+    @objc public var context: SpeechContext
+    
+    private var isActive: Bool = false
+    private var activation = 0
+
+    private var rmsValue: Float = 0.0
+    private var prevSample: Float = 0.0
+    private var sampleWindow: RingBuffer<Float>!
+    private var fftFrame: Array<Float> = []
+    private var fftWindow: Array<Float> = []
+    private var fft: FFT!
+    private var hopLength: Int = 0
+    private var frameWindow: RingBuffer<Float>!
+    private var encodeState: RingBuffer<Float>!
+    private var encodeWindow: RingBuffer<Float>!
+    private var classes: [String] = []
+    
+    // TensorFlowLite models
+    private var filterModel: Interpreter?
+    private var encodeModel: Interpreter?
+    private var detectModel: Interpreter?
+    internal enum Tensors: Int, CaseIterable {
+        case encode
+        case state
+    }
+    
+    
+    private var sampleCollector: Array<Float>?
+    private var fftFrameCollector: String?
+    private var filterCollector: Array<Float>?
+    private var encodeCollector: Array<Float>?
+    
+    
+    /// Initializes a TFLiteKeywordRecognizer instance.
+    ///
+    /// A recognizer is initialized by, and receives `startStreaming` and `stopStreaming` events from, an instance of `SpeechPipeline`.
+    ///
+    /// The TFLiteKeywordRecognizer receives audio data frames to `process` from `AudioController`.
+    /// - Parameters:
+    ///   - configuration: Configuration for the recognizer.
+    ///   - context: Global state for the speech pipeline.
+    @objc public init(_ configuration: SpeechConfiguration, context: SpeechContext) {
+        self.configuration = configuration
+        self.context = context
+        super.init()
+        self.initialize(configuration)
+    }
+    
+    private func initialize(_ c: SpeechConfiguration) {
+        self.rmsValue = c.keywordRMSTarget
+        self.sampleWindow = RingBuffer(c.keywordFFTWindowSize, repeating: 0.0)
+        self.fftFrame = Array(repeating: 0.0, count: c.keywordFFTWindowSize)
+        self.fftWindow = SignalProcessing.fftWindowDispatch(windowType: c.keywordFFTWindowType, windowLength: c.keywordFFTWindowSize)
+        self.fft = FFT(c.keywordFFTWindowSize)
+        self.hopLength = c.keywordFFTHopLength * c.sampleRate / 1000
+        let melLength: Int = c.keywordMelFrameLength * c.sampleRate / 1000 / self.hopLength
+        self.frameWindow = RingBuffer(melLength * c.keywordMelFrameWidth, repeating: 0.0)
+        self.encodeState = RingBuffer(c.keywordEncodeWidth, repeating: 0.0)
+        self.encodeState.fill(0.0) // fill now because the encoded state is used to feed-forward
+        let encodeLength = c.keywordEncodeLength * c.sampleRate / 1000 / self.hopLength
+        self.encodeWindow = RingBuffer(encodeLength * c.keywordEncodeWidth, repeating: -1.0)
+        self.classes = self.configuration.keywords.components(separatedBy: ",")
```

Comment:
the length of the class list should be validated against the detector model output dimension, especially if you insist on keeping that silly default class list configuration